File: R/bicriterion_iterated_local_search_call.R

#' Bicriterion iterated local search heuristic
#'
#' This function implements the bicriterion algorithm BILS for
#' anticlustering by Brusco et al. (2020;
#' <doi:10.1111/bmsp.12186>). The description of their algorithm is
#' given in Section 3 of their paper (in particular, see the
#' Pseudocode in their Figure 2). As of anticlust version 0.8.6, this
#' function also includes some extensions to the BILS algorithm that
#' are implemented through the optional arguments
#' \code{dispersion_distances}, \code{average_diversity},
#' \code{init_partitions}, and \code{return}. If these arguments are
#' not changed, the function performs the "vanilla" BILS as described
#' in Brusco et al.
#' 
#' 
#' @param x The data input. Can be one of two structures: (1) A
#'     feature matrix where rows correspond to elements and columns
#'     correspond to variables (a single numeric variable can be
#'     passed as a vector). (2) An N x N matrix dissimilarity matrix;
#'     can be an object of class \code{dist} (e.g., returned by
#'     \code{\link{dist}} or \code{\link{as.dist}}) or a \code{matrix}
#'     where the entries of the upper and lower triangular matrix
#'     represent pairwise dissimilarities.
#' @param K How many anticlusters should be created. Alternatively:
#'     (a) A vector describing the size of each group, or (b) a vector
#'     of length \code{nrow(x)} describing how elements are assigned
#'     to anticlusters before the optimization starts.
#' @param R The desired number of restarts for the algorithm. By
#'     default, both phases (MBPI + ILS) of the algorithm are
#'     performed once.  See details.
#' @param W Optional argument, a vector of weights defining the
#'     relative importance of dispersion and diversity (0 <= W <=
#'     1). See details.
#' @param Xi Optional argument, specifies probability of swapping
#'     elements during the iterated local search. See examples.
#' @param dispersion_distances A distance matrix used to compute the
#'     dispersion if the dispersion should not be computed on the
#'     basis of argument \code{x}.
#' @param average_diversity Boolean. Compute the diversity not as a
#'     global sum across all pairwise within-group distances, but as
#'     the sum of the average of within-group distances.
#' @param init_partitions A matrix of initial partitions (rows =
#'     partitions; columns = elements) that serve as starting
#'     partitions during the iterations of the first phase of the BILS
#'     (i.e., the MBPI).  If not passed, a new random partition is
#'     generated at the start of each iteration (which is the default
#'     behaviour).
#' @param return Either "paretoset" (default), "best-diversity", 
#'     "best-average-diversity", "best-dispersion". See below.
#'     
#' @details
#'
#' The bicriterion algorithm by Brusco et al. (2020) aims to
#' simultaneously optimize two anticlustering criteria: the
#' \code{\link{diversity_objective}} and the
#' \code{\link{dispersion_objective}}. It returns a list of partitions
#' that approximate the pareto set of efficient solutions across both
#' criteria. By considering both the diversity and dispersion, this
#' algorithm is well-suited for maximizing overall within-group
#' heterogeneity. To select a partition among the approximated pareto
#' set, it is reasonable to plot the objectives for each partition
#' (see Examples).
#'
#' The arguments \code{R}, \code{W} and \code{Xi} are named for
#' consistency with Brusco et al. (2020). The argument \code{K} is
#' used for consistency with other functions in anticlust; Brusco et
#' al. used `G` to denote the number of groups. However, note that
#' \code{K} can not only be used to denote the number of equal-sized
#' groups, but also to specify group sizes, as in
#' \code{\link{anticlustering}}.
#' 
#' This function implements the combined bicriterion algorithm BILS,
#' which consists of two phases: The multistart bicriterion pairwise
#' interchange heuristic (MBPI, which is a local maximum search
#' heuristic similar to \code{method = "local-maximum"} in
#' \code{\link{anticlustering}}), and the iterated local search (ILS),
#' which is an improvement procedure that tries to overcome local
#' optima.  The argument \code{R} denotes the number of restarts of
#' the two phases of the algorithm. If \code{R} has length 1, half of
#' the repetitions perform the first phase MBPI), the other half
#' perform the ILS.  If \code{R} has length 2, the first entry
#' indicates the number of restarts of MBPI the second entry indicates
#' the number of restarts of ILS.  The argument \code{W} denotes the
#' relative weight given to the diversity and dispersion criterion in
#' a given run of the search heuristic. In each run, the a weight is
#' randomly selected from the vector \code{W}. As default values, we
#' use the weights that Brusco et al. used in their analyses. All
#' values in \code{W} have to be in [0, 1]; larger values indicate
#' that diversity is more important, whereas smaller values indicate
#' that dispersion is more important; \code{w = .5} implies the same
#' weight for both criteria. The argument \code{Xi} is the probability
#' that an element is swapped during the iterated local search
#' (specifically, Xi has to be a vector of length 2, denoting the
#' range of a uniform distribution from which the probability of
#' swapping is selected). For \code{Xi}, the default is selected
#' consistent with the analyses by Brusco et al.
#'
#' If the data input \code{x} is a feature matrix (that is: each row
#' is a "case" and each column is a "variable"), a matrix of the
#' Euclidean distances is computed as input to the algorithm. If a
#' different measure of dissimilarity is preferred, you may pass a
#' self-generated dissimilarity matrix via the argument \code{x}.  The
#' argument \code{dispersion_distances} can additionally be used if
#' the dispersion should be computed on the basis of a different
#' distance matrix.
#' 
#' If multiple \code{init_partitions} are given, ensure that each
#' partition (i.e., each row of\code{init_partitions}) has the exact
#' same output of \code{\link{table}}.
#' 
#' @return By default, a \code{matrix} of anticlustering partitions
#'     (i.e., the approximated pareto set). Each row corresponds to a
#'     partition, each column corresponds to an input element. If the
#'     argument \code{return} is set to either "best-diversity", 
#'     "best-average-diversity", or "best-dispersion", it only returns one 
#'     partition (as a vector), that maximizes the respective objective.
#' 
#' @author Martin Breuer \email{M.Breuer@@hhu.de}, Martin Papenberg
#'     \email{martin.papenberg@@hhu.de}
#' 
#' @export
#' 
#' @examples 
#' # Generate some random data
#' M <- 3
#' N <- 80
#' K <- 4
#' data <- matrix(rnorm(N * M), ncol = M)
#'
#' # Perform bicriterion algorithm, use 200 repetitions
#' pareto_set <- bicriterion_anticlustering(data, K = K, R = 200)
#'
#' # Compute objectives for all solutions
#' diversities_pareto <- apply(pareto_set, 1, diversity_objective, x = data)
#' dispersions_pareto <- apply(pareto_set, 1, dispersion_objective, x = data)
#'
#' # Plot the pareto set
#' plot(
#'   diversities_pareto,
#'   dispersions_pareto,
#'   col = "blue",
#'   cex = 2,
#'   pch = as.character(1:NROW(pareto_set))
#' )
#' 
#' # Get some random solutions for comparison
#' rnd_solutions <- t(replicate(n = 200, sample(pareto_set[1, ])))
#' 
#' # Compute objectives for all random solutions
#' diversities_rnd <- apply(rnd_solutions, 1, diversity_objective, x = data)
#' dispersions_rnd <- apply(rnd_solutions, 1, dispersion_objective, x = data)
#' 
#' # Plot random solutions and pareto set. Random solutions are far away 
#' # from the good solutions in pareto set
#' plot(
#'   diversities_rnd, dispersions_rnd, 
#'   col = "red",
#'   xlab = "Diversity",
#'   ylab = "Dispersion",
#'   ylim = c(
#'     min(dispersions_rnd, dispersions_pareto), 
#'     max(dispersions_rnd, dispersions_pareto)
#'   ),
#'   xlim = c(
#'     min(diversities_rnd, diversities_pareto), 
#'     max(diversities_rnd, diversities_pareto)
#'   )
#' )
#' 
#' # Add approximated pareto set from bicriterion algorithm:
#' points(diversities_pareto, dispersions_pareto, col = "blue", cex = 2, pch = 19)
#'
#' @note
#'
#' For technical reasons, the pareto set returned by this function has
#' a limit of 500 partitions. Usually however, the
#' algorithm usually finds much fewer partitions. There is one following exception:
#' We do not recommend to use this method when the input data is
#' one-dimensional where the algorithm may identify too many
#' equivalent partitions causing it to run very slowly (see section 5.6 in 
#' Breuer, 2020).
#' 
#' @references
#' 
#' Brusco, M. J., Cradit, J. D., & Steinley, D. (2020). Combining
#' diversity and dispersion criteria for anticlustering: A bicriterion
#' approach. British Journal of Mathematical and Statistical
#' Psychology, 73, 275-396. https://doi.org/10.1111/bmsp.12186
#' 
#' Breuer (2020). Using anticlustering to maximize diversity and dispersion:
#' Comparing exact and heuristic approaches. Bachelor thesis.
#' 

bicriterion_anticlustering <- function(
  x, K, R = NULL, 
  W = c(0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1, 0.5, 0.99, 0.999, 0.999999),
  Xi = c(0.05, 0.1),
  dispersion_distances = NULL, average_diversity = FALSE, init_partitions = NULL, return = "paretoset") {
  
  input_validation_bicriterion_anticlustering(x, K, R, W, Xi, dispersion_distances, average_diversity, init_partitions, return)

  distances <- convert_to_distances(x) 
  N <- NROW(distances)
  WL <- length(W)
  if (is.null(R)) {
    R <- c(1, 1)
  } else if (length(R) == 1) {
    R <- rep(ceiling(R / 2), 2)
  }
  
  if (is.null(dispersion_distances)) {
    dispersion_distances <- distances
  } else {
    dispersion_distances <- convert_to_distances(dispersion_distances)
  }

  clusters <- initialize_clusters(N, K, NULL) - 1
  
  if (argument_exists(init_partitions)) {
    use_init_partitions <- 1
    init_partitions <- t(apply(init_partitions, 1, to_numeric)) - 1 # ensure correct labeling of partitions
    clusters <- init_partitions[1, ]
  } else {
    init_partitions <- 0
    use_init_partitions <- 0
  }
  
  frequencies <- rep(1, length(unique(clusters)))
  if (average_diversity) {
    frequencies <- table(clusters)
  }
  upper_bound = 500 # limits number of resulting partitions 
  # create empty matrix for results to use in C
  result_matrix = matrix(data = -1, nrow = upper_bound , ncol = N) 
  
  # Call C function
  results <- .C(
    "bicriterion_iterated_local_search_call",
    as.double(distances),
    as.double(dispersion_distances),
    as.integer(N),
    as.integer(R),
    as.integer(upper_bound),
    as.integer(WL),
    as.double(W),
    as.double(Xi),
    as.integer(clusters),
    as.integer(frequencies),
    as.integer(use_init_partitions),
    as.integer(t(init_partitions)),
    result = integer(length(result_matrix)),
    mem_error = as.integer(0),
    PACKAGE = "anticlust" # important to call C
  )

  if (results[["mem_error"]] == 1) {
    stop("Could not allocate enough memory.")
  }
  
  results <- results$result
  
  # remove allocated space that was not needed
  results <- results[results != -1]

  # C returns the list of partitions as one vector, 
  # that we turn back into a matrix
  results <- matrix(results, ncol = N, byrow = TRUE)
  # Order each partition:
  results <- data.frame(t(apply(results, 1, order_cluster_vector)))
  # Remove duplicates
  results <- results[!duplicated(results), ]
  results <- unname(as.matrix(results))
  if (return == "paretoset") {
    return(results)
  }
  if (return == "best-dispersion") {
    best_obj <- which.max(apply(results, 1, dispersion_objective_, dispersion_distances))
    return(results[best_obj, ])
  } else if (return == "best-average-diversity" || average_diversity) {
    # manually compute objectives, apply() does not work here because table() has different results for the different clusterings
    average_diversities <- c()
    for (i in 1:nrow(results)) {
      average_diversities[i] <- weighted_diversity_objective_(distances, results[i, ], table(results[i, ]))
    }
    return(results[which.max(average_diversities), ])
  } else if (return == "best-diversity") {
    best_obj <- which.max(apply(results, 1, diversity_objective_, distances))
    return(results[best_obj, ])
  } 
}

input_validation_bicriterion_anticlustering <- function(
    x, K, R, W, Xi, dispersion_distances, 
    average_diversity, init_partitions, return) {

  input_validation_anticlustering(
    x, K, objective = "diversity", method = "brusco", 
    preclustering = FALSE, categories = NULL,
    repetitions = 1, standardize = FALSE
  )
  
  x <- convert_to_distances(x)
  N <- nrow(x)
  
  validate_input(
    return, "return", objmode = "character", len = 1,
    input_set = c("paretoset", "best-diversity", "best-average-diversity", "best-dispersion"), 
    not_na = TRUE, 
    not_function = TRUE
  )
  
  validate_input(
    average_diversity, "average_diversity", objmode = "logical", len = 1,
    input_set = c(TRUE, FALSE), 
    not_na = TRUE, 
    not_function = TRUE
  )
  
  if (average_diversity && return == "best-diversity") {
    stop("Cannot use 'return = best-diversity' with 'average_diversity = TRUE' (probably you want to use 'return = best-average-diversity').")
  }
  if (!average_diversity && return == "best-average-diversity") {
    stop("Cannot use 'return = best-average-diversity' with 'average_diversity = FALSE' (probably you want to use 'return = best-diversity').")
  }

  if (argument_exists(dispersion_distances)) {
    if (any(dim(dispersion_distances) != dim(x))) {
      stop("The dimensions of argument 'x' and 'dispersion_distances' do not match.")
    } 
  }

  checkweights(W)
  checkneighborhood(Xi)
  
  if (argument_exists(R)) {
    validate_input(R, "R", must_be_integer = TRUE, not_na = TRUE, not_function = TRUE, greater_than = -1)
    if (!length(R) %in% 1:2) {
      stop("Argument 'R' must have length 1 or 2.")
    }
  }
  
  if (argument_exists(init_partitions)) {
    if (length(R) != 2) {
      stop("If 'init_partitions' is used, specify argument R as vector of length 2.")
    }
    if (nrow(init_partitions) != R[1]) {
      stop("The number of repetitions (argument 'R') and the number of partitions (argument 'init_partitions') do not match.")
    } 
    if (ncol(init_partitions) != N) {
      stop("The number of columns in 'init_partitions' does not match the number of cases in 'x'.")
    }
  }
}


#verify input
#############
checkweights <- function(W){
  validate_input(W, "W", not_na = TRUE, not_function = TRUE, objmode = "numeric")
  for(i in W){
    if(i < 0 | i > 1){
      stop("All weights (argument 'W') must be percentages. Only values between 0 and 1 are allowed.")
    }
  }  
}

checkneighborhood <- function(Xi) {
  validate_input(Xi, "Xi", not_na = TRUE, not_function = TRUE, len = 2, objmode = "numeric")
  for (i in Xi) {
    if (i < 0 | i > 1) {
      stop("Both neighorhood indexes must be a percentage. Only values between 0 and 1 are allowed.")
    }
  } 
  if (Xi[1] > Xi[2]) {
    stop("First neighborhood percentage needs to be smaller than the second.")
  }
}

File: src/bicriterion_iterated_local_search_call.c
#include <stdlib.h> 
#include <stdio.h>
#include <math.h>
#include <stdbool.h>
#include <string.h>
#include "header.h"
#include <R.h>

//datastructure to store a linked-list of partitions
struct Pareto_element { 
  double diversity; 
  double dispersion;
  int *partition;
  struct Pareto_element *next;
};

//receive data from r, call bils algorithm, save results for r
void bicriterion_iterated_local_search_call(double *distances, int *N, int *R, 
                                            int *upper_bound, int *WL, double *W, double *Xi, 
                                            int *partition,
                                            int *result,
                                            int *mem_error
                                           ) {
  
  const size_t n = *N; // number of elements
  const size_t r = *R; // number of restarts
  const size_t u = *upper_bound; //max. length of result-list
  const size_t wl = *WL; // length of possible weights
  
  int distance_ptr[n]; // indexing columns
  double distance_pts[n][n];
  
  // Column offsets (to convert one-dimensional array to Row/Col major)
  for (size_t i = 0; i < n; i++) {
     distance_ptr[i] = i * n;
  }
  
  // Reconstruct the data points as N x N matrix
  for (size_t i = 0; i < n; i++) {
    for (size_t j = 0; j < n; j++) {
      distance_pts[i][j] = distances[distance_ptr[j]++];
    }
  }
  
  double weights[wl];
  for (size_t i = 0; i < wl; i++){
    weights[i] = W[i];
  }
  
  double neighbor_percent[2];
  neighbor_percent[0] = Xi[0];
  neighbor_percent[1] = Xi[1];
  
  //divide restarts for both parts of the algorithm equally
  size_t half_restarts = r/2 + (r%2);
  
  struct Pareto_element* head = multistart_bicriterion_pairwise_interchange(n, 
            distance_pts, 
            half_restarts, 
            wl, 
            weights, 
            partition
  );
  if (head == NULL) {
    *mem_error = 1; // return after memory allocation error
    return;
  }
  head = bicriterion_iterated_local_search(head, n, distance_pts, half_restarts, wl, weights, neighbor_percent);
  if (head == NULL) {
    *mem_error = 1; // return after memory allocation error
    return;
  }
  
  //fill result with -1 to identify the last partition in the matrix
  for (size_t i = 0; i < n*u; i++){
    result[i] = -1;
  }
  //update result with the solutions from the linked list
  size_t position = 0;
  struct Pareto_element* tmp = head; // for freeing memory lateron
  while (head != NULL && position != n*u) {
    for (size_t i = 0; i < n; i++){
      result[position] = head->partition[i];
      position++;
    }
    head = head->next;
  } 
  free_pareto_set(tmp);
  return;
}

// function to free all memory allocated by the pareto set
void free_pareto_set(struct Pareto_element* head) {
  struct Pareto_element* tmp; // for freeing memory
  while (head != NULL) {
    tmp = head;
    head = head->next;
    free(tmp->partition);
    free(tmp);
  }
  return;
}

// returns the HEAD to a pareto set (linked list); if it returns NULL, a memory allocation error occurred
struct Pareto_element* multistart_bicriterion_pairwise_interchange(
    size_t N, 
    double matrix[N][N], 
    size_t R, 
    size_t WL, 
    double weights[WL], 
    int *partition) {
  
  struct Pareto_element* head = NULL; // head pointing on the later linked list(paretoset)
  
  for (size_t a = 0; a < R; a++){
    if (a > 0) {
        shuffle_permutation(N, partition);
    }
    double div_weight = sample(WL, weights); 
    double dis_weight = 1 - div_weight;
    double diversity = get_diversity(N, partition, matrix);
    double save_diversity = diversity;
    double dispersion = get_dispersion(N, partition, matrix);
    double save_dispersion = dispersion;
    double max_bicriterion = div_weight*diversity + dis_weight*dispersion;
    bool Flag = false;
    while(!Flag){
      Flag = true;
      for (size_t i = 0; i < N-1; i++){
        for (size_t j = i+1; j < N; j++){
          int g = partition[i];
          int h = partition[j];
          if(g != h){
            cluster_swap(i, j, partition);
            double current_diversity = get_diversity_fast(save_diversity, i, j, N, partition, matrix);
            double current_dispersion = get_dispersion_fast(save_dispersion , i, j, N, partition, matrix);
            if (update_pareto(&head, N, partition,current_diversity, current_dispersion) == 1) {
                free_pareto_set(head); // free all memory
                return NULL;
            }
            double current_bicriterion = div_weight*current_diversity + dis_weight*current_dispersion;
            if(current_bicriterion > max_bicriterion){
              save_diversity = current_diversity;
              save_dispersion = current_dispersion;
              max_bicriterion = current_bicriterion;
              Flag = false;
            }else{
              cluster_swap(i, j, partition);
            }
          }
        }
      }
    }
  }
  return (head);
}  

// returns the HEAD to a pareto set (linked list); if it returns NULL, a memory allocation error occurred
struct Pareto_element* bicriterion_iterated_local_search(
    struct Pareto_element* head, size_t N, double matrix[N][N], size_t R, 
    size_t WL, double weights[WL], double neighbor_percent[2]){
  
  for (size_t a = 0; a < R; a++){
    double div_weight = sample(10, weights); 
    double dis_weight = 1 - div_weight;
    double neighborhood_size = uni_rnd_number_range(neighbor_percent[0], neighbor_percent[1]);
    int* partition = (int*) malloc(sizeof(int) * N);
    linked_list_sample(head, N, partition);
    for (size_t i = 0; i < N-1; i++){
      for (size_t j = i + 1; j < N; j++){
        int g = partition[i];
        int h = partition[j];
        if (g != h){
          double random = uni_rnd_number_range(0,1);
          if(random < neighborhood_size){
            cluster_swap(i, j, partition);
          }
        }
      }
    }
    double diversity = get_diversity(N, partition, matrix);
    double save_diversity = diversity;
    double dispersion = get_dispersion(N, partition, matrix);
    double save_dispersion = dispersion;
    double max_bicriterion = div_weight*diversity + dis_weight*dispersion;
    bool Flag = false;
    while(!Flag){
      Flag = true;
      for (size_t i = 0; i < N-1; i++){
        for (size_t j = i+1; j < N; j++){
          int g = partition[i];
          int h = partition[j];
          if(g != h){
            cluster_swap(i, j, partition);
            double current_diversity = get_diversity_fast(save_diversity, i, j, N, partition, matrix);
            double current_dispersion = get_dispersion_fast(save_dispersion , i, j, N, partition, matrix);
            if (update_pareto(&head, N, partition, current_diversity, current_dispersion) == 1) {
                free_pareto_set(head); // free all memory
                free(partition);
                return NULL;
            }
            double current_bicriterion = div_weight*current_diversity + dis_weight*current_dispersion;
            if(current_bicriterion > max_bicriterion){
              save_diversity = current_diversity;
              save_dispersion = current_dispersion;
              max_bicriterion = current_bicriterion;
              Flag = false;
            }else{
              cluster_swap(i, j, partition);
            }
          }
        }
      }
    }
    free(partition);
  }
  return (head);
}     

//sample functin from R. Get one random element from an array
double sample(size_t array_size, double array[array_size]) {
  int max = (int) array_size;
  int r = random_integer(0, max-1);
  return(array[r]);
}

double get_diversity(size_t N, int* partition, double matrix[N][N]){
  
  double sum = 0;
  
  for (size_t i = 0; i < N-1; i++){
    for (size_t j = i+1; j < N; j++){
      if (partition[i] == partition[j]){
        sum = sum + matrix[i][j];
      }
    }
  }
  
  return(sum);  
}


double get_dispersion(size_t N, int* partition, double matrix[N][N]){
  
  double min = INFINITY;
  double distance;
  
  for (size_t i = 0; i < N-1; i++){
    for (size_t j = i+1; j < N; j++){
      if (partition[i] == partition[j]){
        distance = matrix[i][j];
        if (distance < min){
          min = distance;
        }
      }
    }
  }
  
  return(min);
}


// Fisher-Yates shuffle algorithm for shuffling ermutations
void shuffle_permutation(int N, int *permutation) {
    for (int i = 0; i <= N-2; i++) {
        int j = random_integer(0, i);
        cluster_swap(i, j, permutation);
    }
}

void cluster_swap(size_t i, size_t j, int* partition){
  int save = partition[i];
  partition[i] = partition[j];
  partition[j] = save;
}


int update_pareto(struct Pareto_element** head_ref, size_t N, int* partition, double diversity, double dispersion){
  
  if(*head_ref == NULL){
    if (push(head_ref, diversity, dispersion, N, partition) == 1) {
        return 1;
    }
    
  }
  if (!(paretodominated(*head_ref, diversity, dispersion))) {
      delete_outdated(head_ref, diversity, dispersion);
      if (push(head_ref, diversity, dispersion, N, partition) == 1) {
        return 1;
      }
  }
  return 0;
}

bool paretodominated(struct Pareto_element* head, double diversity, double dispersion){
  
  while(head != NULL){
    if((head->diversity >= diversity && head->dispersion > dispersion) || (head->diversity > diversity && head->dispersion >= dispersion)){
      return (true);
    }
    head = head->next;
  }
  
  return (false);
}

// returns 1 if a memory allocation error occurs, 0 otherwise
int push(struct Pareto_element** head_ref, double diversity, double dispersion, size_t N, int* partition) { 
  
        /* 1. allocate node */
        struct Pareto_element* new_node = (struct Pareto_element*) malloc(sizeof(struct Pareto_element)); 
        if (new_node == NULL) {
            return 1;
        }
        
        /* 2. put in the data  */
        new_node->diversity  = diversity;
        new_node->dispersion = dispersion;
        new_node->partition = (int*) malloc(sizeof(int) * N);  // vllt memcpy anstatt schleife?
        if (new_node->partition == NULL) {
            free(new_node);
            return 1;
        }
        for(int b = 0;b<N;b++){ 
            new_node->partition[b]  = partition[b];
        }
        
        /* 3. Make next of new node as head */
        new_node->next = (*head_ref); 
            
        /* 4. move the head to point to the new node */
        (*head_ref)    = new_node; 
        return 0;
}  


void delete_outdated(struct Pareto_element** head_ref, double diversity, double dispersion){
  
        struct Pareto_element* temp = *head_ref;
        struct Pareto_element* prev = *head_ref;
        struct Pareto_element* safe = NULL;
        
        while(temp != NULL && ((diversity >= temp->diversity && dispersion > temp->dispersion) || 
            (diversity > temp->diversity && dispersion >= temp->dispersion))){
            *head_ref = temp->next; 
            prev = temp->next;
            free(temp->partition);
            free(temp); 
            temp = prev;
        }
        if(temp != NULL){
            temp = temp->next;
        }
        
        while(temp != NULL){
            if((diversity >= temp->diversity && dispersion > temp->dispersion) || 
                (diversity > temp->diversity && dispersion >= temp->dispersion)){
            prev->next = temp->next;
            safe = temp->next;
            free(temp->partition);
            free(temp);
            temp = safe;
            }else{
            prev = prev->next;
            temp = temp->next;
            }
        }
} 


// selects a random partition
void linked_list_sample(struct Pareto_element* head, size_t N, int* partition){
  
  int count = linked_list_length(head);
  int r = random_integer(0, count-1);
  struct Pareto_element* current = head; 
  
  for (int i = 0; i < r; i++){
    current = current->next;
  }
  
  for (int i = 0; i < N; i++){
    partition[i] = current->partition[i];
  }
  
}

int linked_list_length(struct Pareto_element* head) {
  
  int count = 0;   
  struct Pareto_element* current = head;  
  while (current != NULL) 
  { 
    count++; 
    current = current->next; 
  } 
  return count; 
} 


double get_diversity_fast(double diversity, int x, int y, size_t N, int* partition, double matrix[N][N]){
  
  int cluster_x = partition[x];
  int cluster_y = partition[y];
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_y && i != x && i != y){
      diversity -= matrix[i][x];
    }
  }
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_x && i != x){
      diversity += matrix[i][x];
    }
  }
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_x && i != x && i != y){
      diversity -= matrix[i][y];
    }
  }
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_y && i != y){
      diversity += matrix[i][y];
    }
  }
  
  return(diversity);  
}


double get_dispersion_fast(double dispersion, int x,int y, size_t N, int* partition, double matrix[N][N]){
  
  int cluster_x = partition[x];
  int cluster_y = partition[y];
  bool before = false;
  bool after = false;
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_y  && i != x && i != y){
      if(dispersion == matrix[i][x]){
        before = true;
        break;
      }
    }
  }
  
  for(int i = 0; i < N && (!before); i++){
    if(partition[i] == cluster_x && i != x && i != y){
      if(dispersion == matrix[i][y]){
        before = true;
        break;
      }
    }
  }
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_x && i != x){
      if(dispersion >= matrix[i][x]){
        dispersion = matrix[i][x];
        after = true;
      }
    }
  }
  
  for(int i = 0; i < N; i++){
    if(partition[i] == cluster_y && i != y){
      if(dispersion >= matrix[i][y]){
        dispersion = matrix[i][y];
        after = true;
      }
    }
  }
  
  
  if(!before && !after){
    return(dispersion);
  }
  
  if(!before && after){
    return(dispersion);
  }
  
  if(before && after){
    return(dispersion);
  }
  
  if(before && !after){
    dispersion = get_dispersion(N, partition, matrix);
    return(dispersion);
  }
  
  return(dispersion);
}

// generate a random uniform number in range
double uni_rnd_number_range(double min, double max) {
  double number = uniform_rnd_number();
  return (min + number * (max - min));
}

// Generate a random integer in given range
int random_integer(int min, int max) {
  int integer = (int) floor(uniform_rnd_number() * (max - min + 1)) + min;
  return integer;
}

double uniform_rnd_number() {
  GetRNGstate();
  double my_number = unif_rand();
  PutRNGstate();
  return my_number;
}
